src/samplers/libcall/mod.rs

```diff
@@ -0,0 +1,278 @@
+
+use std::sync::{Arc, Mutex};
+use std::time::Instant;
+
+use crate::common::bpf::BPF;
+use crate::config::SamplerConfig;
+use crate::samplers::{Common, Sampler};
+
+#[cfg(feature = "bpf")]
+use crate::common::bpf::bpf_hash_char_to_map;
+
+#[cfg(feature = "bpf")]
+use rustcommon_metrics::Statistic;
+
+mod config;
+mod stat;
+
+pub use config::{LibCallConfig, LibFileMap, LibSearchMap};
+pub use stat::LibCallStatistic;
+
+use std::path::Path;
+
+#[allow(dead_code)]
+pub struct LibCall {
+    bpf: Option<Arc<Mutex<BPF>>>,
+    bpf_last: Arc<Mutex<Instant>>,
+    common: Common,
+    statistics: Vec<LibCallStatistic>,
+    lib_files: LibFileMap,
+    lib_search: LibSearchMap,
+}
+
+#[cfg(feature = "bpf")]
+const PROBE_PRELUDE: &str = r#"
+#include <uapi/linux/ptrace.h>
+struct key_t {
+    char c[80];
+};
+BPF_HASH(counts, struct key_t);
+
+"#;
+
+macro_rules! probe_template {
+    () => {
+        r#"
+
+int probe_{}(void *ctx) {{
+    struct key_t key = {{.c = "{}"}};
+    u64 zero = 0, *val;
+    val = counts.lookup_or_init(&key, &zero);
+    (*val)++;
+    return 0;
+}}
+"#
+    };
+}
+
+#[allow(dead_code)]
+fn path_match(lib_name: &str, path: &Path) -> bool {
+    if let Some(file_name) = path.file_name() {
+        if let Some(file_str) = file_name.to_str() {
+            let parts: Vec<&str> = file_str.split('.').collect();
+            if parts.len() < 2 {
+                return false;
+            }
+            let mut stem_str: String = parts[0].to_string();
+            let mut ext_str: String = parts[1].to_string();
+            let end_index = parts.len() - 1;
+            if parts[end_index] == "so" {
+                stem_str = parts[..end_index].join(".");
+                ext_str = "so".into();
+            }
+            let to_test = match stem_str.starts_with("lib") {
+                true => stem_str[3..].into(),
+                false => stem_str,
+            };
+            return to_test.eq(lib_name) && "so".eq(&ext_str[..]);
+        }
+    }
+    false
+}
+
+impl LibCall {
+    fn init_bpf(&mut self) -> Result<(), anyhow::Error> {
+        // The bpf source code generated by probe_template
+        let mut bpf_probes = String::new();
+
+        // The list of probes that have been found.
+        let mut found_probes: Vec<(String, &str, &str)> = Vec::new();
+
+        // Add probes that are linked to specific files in the config
+        for (lib, func_map) in &self.lib_files {
+            for (file, funcs) in func_map {
+                for func in funcs.iter() {
+                    bpf_probes.push_str(&format!(
+                        probe_template!(),
+                        found_probes.len(),
+                        format!("{}/{}", lib, func)
+                    ));
+                    found_probes.push((file.clone(), lib, func));
+                }
+            }
+        }
+
+        // Add probes by searching the default paths.
+        let default_paths: Vec<String> = vec![
```

Comment:
Should the default path variable be promoted to the top of the module as a constant?

Reply:
Yeah, I think that's a good idea. I moved these paths to a const.